File: modules/expo-tor/android/src/main/java/expo/modules/tor/ExpoTorModule.kt
package expo.modules.tor

import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.ServiceConnection
import android.os.Build
import android.os.IBinder
import android.util.Log
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.URL
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.torproject.jni.TorService

class ExpoTorModule : Module() {
  private var torService: TorService? = null
  private var isBound = false
  private var bindRequested = false  // Track if we requested a bind
  private var isReceiverRegistered = false
  private val TAG = "ExpoTorModule"
  private val moduleScope = CoroutineScope(Dispatchers.Main + Job())
  private var connectionCheckJob: Job? = null

  private val torServiceConnection =
          object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
              val binder = service as? TorService.LocalBinder
              torService = binder?.getService()
              isBound = true
              Log.d(TAG, "✅ Tor Service Connected")

              // Wait for Tor Control Connection using coroutines
              connectionCheckJob?.cancel() // Cancel any existing job
              connectionCheckJob = moduleScope.launch(Dispatchers.IO) {
                repeat(20) { attempt ->
                  val connection = torService?.getTorControlConnection()
                  Log.d(TAG, "🔍 Checking Tor control connection (attempt ${attempt + 1}/20): ${if (connection != null) "Connected" else "Not connected"}")
                  if (connection != null) {
                    withContext(Dispatchers.Main) {
                      Log.d(TAG, "✅ Tor Control Connection established!")
                      sendEvent("onTorConnected", mapOf("connected" to true))
                    }
                    return@launch
                  }
                  delay(500)
                }
                Log.w(TAG, "⚠️ Tor Control Connection not established after 10 seconds")
              }
            }

            override fun onServiceDisconnected(name: ComponentName?) {
              connectionCheckJob?.cancel()
              connectionCheckJob = null
              torService = null
              isBound = false
              bindRequested = false
              Log.d(TAG, "🔴 Tor Service Disconnected")
              sendEvent("onTorDisconnected", mapOf("connected" to false))
              sendEvent("onTorStatus", mapOf("status" to "OFF"))
            }
          }

  private val statusReceiver =
          object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
              val status = intent?.getStringExtra(TorService.EXTRA_STATUS)
              Log.d(TAG, "📡 Tor Status Update: $status")
              sendEvent("onTorStatus", mapOf("status" to status))
            }
          }

  override fun definition() = ModuleDefinition {
    Name("ExpoTor")

    // Defines event names that the module can send to JavaScript.
    Events("onTorStatus", "onTorConnected", "onTorDisconnected", "onTorError")

    // Cleanup on module destroy
    OnDestroy {
      try {
        // Cancel any pending coroutine jobs
        connectionCheckJob?.cancel()
        connectionCheckJob = null
        Log.d(TAG, "Coroutine jobs cancelled")

        val context = appContext.reactContext

        if (context != null && isReceiverRegistered) {
          try {
            context.unregisterReceiver(statusReceiver)
            isReceiverRegistered = false
            Log.d(TAG, "BroadcastReceiver unregistered in OnDestroy")
          } catch (e: Exception) {
            Log.w(TAG, "Receiver cleanup in OnDestroy", e)
          }
        }

        if (context != null && (isBound || bindRequested)) {
          try {
            context.unbindService(torServiceConnection)
            isBound = false
            bindRequested = false
            Log.d(TAG, "Service unbound in OnDestroy")
          } catch (e: Exception) {
            Log.w(TAG, "Error unbinding in OnDestroy", e)
          }
        }

        torService = null
      } catch (e: Exception) {
        Log.e(TAG, "Error in OnDestroy cleanup", e)
      }
    }

    // Start Tor Service
    AsyncFunction("startTor") { promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("Context is null")

        Log.d(TAG, "Starting Tor service...")

        // Register status receiver (only if not already registered)
        if (!isReceiverRegistered) {
          val filter = IntentFilter(TorService.ACTION_STATUS)
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            context.registerReceiver(statusReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
          } else {
            context.registerReceiver(statusReceiver, filter)
          }
          isReceiverRegistered = true
          Log.d(TAG, "BroadcastReceiver registered")
        }

        // Bind to TorService (BIND_AUTO_CREATE will start the service if needed)
        val intent = Intent(context, TorService::class.java)
        val bound = context.bindService(intent, torServiceConnection, Context.BIND_AUTO_CREATE)
        bindRequested = bound
        Log.d(TAG, "🚀 Service bind requested (auto-create), result: $bound")

        if (!bound) {
          promise.reject("TOR_START_ERROR", "Failed to bind to Tor service", null)
          return@AsyncFunction
        }

        promise.resolve(mapOf("success" to true, "message" to "Tor service starting..."))
      } catch (e: Exception) {
        Log.e(TAG, "Error starting Tor", e)
        promise.reject("TOR_START_ERROR", e.message, e)
      }
    }

    // Stop Tor Service
    AsyncFunction("stopTor") { promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("Context is null")

        Log.d(TAG, "🛑 Stopping Tor service...")

        // Cancel any pending connection check jobs
        connectionCheckJob?.cancel()
        connectionCheckJob = null
        Log.d(TAG, "🛑 Connection check job cancelled")

        // Unbind from service (if we requested a bind, we must unbind)
        if (bindRequested || isBound) {
          try {
            Log.d(TAG, "🛑 Attempting to unbind service (bindRequested: $bindRequested, isBound: $isBound)")
            context.unbindService(torServiceConnection)
            isBound = false
            bindRequested = false
            Log.d(TAG, "🛑 Service unbound successfully")
          } catch (e: Exception) {
            Log.w(TAG, "⚠️ Error unbinding service", e)
            // Reset flags even if unbind failed
            isBound = false
            bindRequested = false
          }
        } else {
          Log.d(TAG, "⚠️ No service to unbind (bindRequested: $bindRequested, isBound: $isBound)")
        }

        // Unregister receiver
        if (isReceiverRegistered) {
          try {
            context.unregisterReceiver(statusReceiver)
            isReceiverRegistered = false
            Log.d(TAG, "🛑 BroadcastReceiver unregistered")
          } catch (e: Exception) {
            Log.w(TAG, "⚠️ Receiver not registered", e)
          }
        }

        torService = null
        Log.d(TAG, "✅ Tor service stopped")

        promise.resolve(mapOf("success" to true, "message" to "Tor service stopped"))
      } catch (e: Exception) {
        Log.e(TAG, "❌ Error stopping Tor", e)
        promise.reject("TOR_STOP_ERROR", "Failed to stop Tor service", e)
      }
    }

    // Get Tor Status
    Function("getTorStatus") { TorService.currentStatus }

    // Get SOCKS Port
    Function("getSocksPort") { torService?.getSocksPort() ?: -1 }

    // Get HTTP Tunnel Port
    Function("getHttpTunnelPort") { torService?.getHttpTunnelPort() ?: -1 }

    // Get Tor Info
    AsyncFunction("getTorInfo") { key: String, promise: Promise ->
      try {
        val info = torService?.getInfo(key)
        if (info != null) {
          promise.resolve(info)
        } else {
          promise.reject("TOR_NOT_READY", "Could not get info for key: $key. Tor may not be ready.", null)
        }
      } catch (e: Exception) {
        Log.e(TAG, "Error getting Tor info", e)
        promise.reject("TOR_CONFIGURATION_ERROR", "Failed to retrieve Tor information", e)
      }
    }

    // Check if Tor is connected
    Function("isTorConnected") { isBound && torService?.getTorControlConnection() != null }

    // Make HTTP request through Tor
    AsyncFunction("makeRequest") { url: String, options: Map<String, Any>?, promise: Promise ->
      try {
        val socksPort = torService?.getSocksPort() ?: -1
        if (socksPort == -1) {
          promise.reject("TOR_NOT_READY", "Tor is not connected yet", null)
          return@AsyncFunction
        }

        // Create SOCKS proxy
        val proxy = Proxy(Proxy.Type.SOCKS, InetSocketAddress("127.0.0.1", socksPort))

        // Create OkHttp client with proxy
        val client =
                OkHttpClient.Builder()
                        .proxy(proxy)
                        .connectTimeout(30, TimeUnit.SECONDS)
                        .readTimeout(30, TimeUnit.SECONDS)
                        .writeTimeout(30, TimeUnit.SECONDS)
                        .build()

        // Parse options
        val method = options?.get("method") as? String ?: "GET"
        val headers = options?.get("headers") as? Map<String, String> ?: emptyMap()
        val body = options?.get("body") as? String

        // Build request
        val requestBuilder = Request.Builder().url(url)

        // Add headers
        headers.forEach { (key, value) -> requestBuilder.addHeader(key, value) }

        // Add body if present
        if (body != null && (method == "POST" || method == "PUT" || method == "PATCH")) {
          val contentType = headers["Content-Type"] ?: "application/json"
          requestBuilder.method(method, body.toRequestBody(contentType.toMediaType()))
        } else {
          requestBuilder.method(method, null)
        }

        val request = requestBuilder.build()

        // Execute request in background using Coroutines
        moduleScope.launch(Dispatchers.IO) {
          try {
            client.newCall(request).execute().use { response ->
              val responseBody = response.body?.string() ?: ""
              val responseHeaders = mutableMapOf<String, String>()
              response.headers.forEach { (name, value) -> responseHeaders[name] = value }

              withContext(Dispatchers.Main) {
                promise.resolve(
                  mapOf(
                    "status" to response.code,
                    "statusText" to response.message,
                    "headers" to responseHeaders,
                    "data" to responseBody,
                    "url" to url
                  )
                )
              }
            }
          } catch (e: Exception) {
            Log.e(TAG, "HTTP request error", e)
            withContext(Dispatchers.Main) {
              promise.reject("HTTP_REQUEST_ERROR", "HTTP request failed: ${e.message}", e)
            }
          }
        }
      } catch (e: Exception) {
        Log.e(TAG, "Request setup error", e)
        promise.reject("REQUEST_SETUP_ERROR", "Failed to setup HTTP request: ${e.message}", e)
      }
    }

    // Enables the module to be used as a native view.
    View(ExpoTorView::class) {
      // Defines a setter for the `url` prop.
      Prop("url") { view: ExpoTorView, url: URL -> view.webView.loadUrl(url.toString()) }
      // Defines an event that the view can send to JavaScript.
      Events("onLoad")
    }
  }
}

File: modules/expo-tor/ios-old/ExpoTorModule.swift
import ExpoModulesCore
@preconcurrency import Tor

/**
 * ExpoTorModule - iOS Implementation
 *
 * This module provides complete Tor network functionality for iOS using the Tor.framework
 * from iCepa (https://github.com/iCepa/Tor.framework).
 *
 * Key Features:
 * - Background Tor thread management using TORThread
 * - SOCKS5 proxy configuration for network requests
 * - Event-based status updates (onTorStatus, onTorConnected, onTorDisconnected, onTorError)
 * - HTTP request proxy through Tor using URLSession with SOCKS proxy
 * - Complete API parity with Android implementation
 *
 * Architecture:
 * - TORThread runs Tor in background (non-blocking)
 * - TORController manages Tor control connection
 * - URLSession configured with SOCKS proxy for makeRequest
 * - All async operations handled on background threads
 */

public class ExpoTorModule: Module {

  // MARK: - Properties

  /// Tor thread instance that runs Tor in the background
  private var torThread: TorThread?

  /// Tor controller for communicating with the Tor process
  private var torController: TorController?

  /// Configuration for Tor
  private var torConfiguration: TorConfiguration?

  /// Current connection status
  private var isConnected: Bool = false

  /// Current Tor status string
  private var currentStatus: String = "STOPPED"

  /// SOCKS port for proxy connections (default 9050, but dynamically assigned)
  private var socksPort: Int = 9050

  /// HTTP tunnel port (control port)
  private var httpTunnelPort: Int = 9051

  /// Queue for background operations
  private let torQueue = DispatchQueue(label: "expo.tor.background", qos: .userInitiated)

  // MARK: - Module Definition

  public func definition() -> ModuleDefinition {
    Name("ExpoTor")

    // Define event names that match Android implementation exactly
    Events("onTorStatus", "onTorConnected", "onTorDisconnected", "onTorError")

    // MARK: - Start Tor
    /**
     * Starts the Tor service in background
     * Returns: Promise<{ success: boolean, message: string }>
     */
    AsyncFunction("startTor") { (promise: Promise) in
      self.torQueue.async {
        do {
          // Check if already running
          if self.torThread != nil {
            DispatchQueue.main.async {
              promise.resolve([
                "success": true,
                "message": "Tor is already running"
              ])
            }
            return
          }

          // Initialize Tor configuration
          try self.initializeTorConfiguration()

          // Create and start Tor thread
          guard let config = self.torConfiguration else {
            throw NSError(domain: "ExpoTor", code: 1, userInfo: [
              NSLocalizedDescriptionKey: "Failed to create Tor configuration"
            ])
          }

          self.torThread = TorThread(configuration: config)

          // Start Tor thread
          self.torThread?.start()

          self.currentStatus = "STARTING"
          self.sendEvent("onTorStatus", ["status": "STARTING"])

          // Wait for Tor to start and establish control connection
          self.establishControlConnection { success in
            if success {
              DispatchQueue.main.async {
                promise.resolve([
                  "success": true,
                  "message": "Tor service starting..."
                ])
              }
            } else {
              DispatchQueue.main.async {
                promise.reject("TOR_START_ERROR", "Failed to establish Tor control connection")
              }
            }
          }

        } catch {
          DispatchQueue.main.async {
            promise.reject("TOR_START_ERROR", error.localizedDescription)
          }
        }
      }
    }

    // MARK: - Stop Tor
    /**
     * Stops the Tor service
     * Returns: Promise<{ success: boolean, message: string }>
     */
    AsyncFunction("stopTor") { (promise: Promise) in
      self.torQueue.async {
        // Disconnect controller
        if let controller = self.torController, controller.isConnected {
          controller.disconnect()
        }

        // Cancel Tor thread
        self.torThread?.cancel()
        self.torThread = nil
        self.torController = nil

        self.isConnected = false
        self.currentStatus = "STOPPED"

        self.sendEvent("onTorDisconnected", ["connected": false])

        DispatchQueue.main.async {
          promise.resolve([
            "success": true,
            "message": "Tor service stopped"
          ])
        }
      }
    }

    // MARK: - Get Tor Status
    /**
     * Returns current Tor status as string
     * Synchronous function
     */
    Function("getTorStatus") { () -> String in
      return self.currentStatus
    }

    // MARK: - Get SOCKS Port
    /**
     * Returns the SOCKS proxy port number
     * Returns -1 if not available
     * Synchronous function
     */
    Function("getSocksPort") { () -> Int in
      return self.socksPort
    }

    // MARK: - Get HTTP Tunnel Port
    /**
     * Returns the HTTP tunnel (control) port number
     * Returns -1 if not available
     * Synchronous function
     */
    Function("getHttpTunnelPort") { () -> Int in
      return self.httpTunnelPort
    }

    // MARK: - Get Tor Info
    /**
     * Gets information from Tor using GETINFO command
     * @param key: The info key to query (e.g., "version", "status/bootstrap-phase")
     * Returns: Promise<string>
     */
    AsyncFunction("getTorInfo") { (key: String, promise: Promise) in
      self.torQueue.async {
        guard let controller = self.torController, controller.isConnected else {
          DispatchQueue.main.async {
            promise.reject("TOR_INFO_ERROR", "Tor controller not connected")
          }
          return
        }

        // Use GETINFO command to retrieve information
        // Note: getInfoForKeys returns [String] array
        controller.getInfoForKeys([key]) { response in
          if let value = response.first {
            DispatchQueue.main.async {
              promise.resolve(value)
            }
          } else {
            DispatchQueue.main.async {
              promise.reject("TOR_INFO_ERROR", "Could not get info for key: \(key)")
            }
          }
        }
      }
    }

    // MARK: - Is Tor Connected
    /**
     * Checks if Tor is connected and ready
     * Synchronous function
     */
    Function("isTorConnected") { () -> Bool in
      return self.isConnected && self.torController?.isConnected == true
    }

    // MARK: - Make Request
    /**
     * Makes an HTTP request through the Tor SOCKS proxy
     * @param url: The URL to request
     * @param options: Optional request options (method, headers, body)
     * Returns: Promise<TorRequestResponse>
     */
    AsyncFunction("makeRequest") { (url: String, options: [String: Any]?, promise: Promise) in
      self.torQueue.async {
        // Check if Tor is ready
        guard self.isConnected, self.socksPort > 0 else {
          DispatchQueue.main.async {
            promise.reject("TOR_NOT_READY", "Tor is not connected yet")
          }
          return
        }

        // Parse options
        let method = options?["method"] as? String ?? "GET"
        let headers = options?["headers"] as? [String: String] ?? [:]
        let body = options?["body"] as? String

        // Create URL
        guard let requestUrl = URL(string: url) else {
          DispatchQueue.main.async {
            promise.reject("INVALID_URL", "Invalid URL: \(url)")
          }
          return
        }

        // Configure URLSession with SOCKS proxy
        // Using string keys for iOS compatibility
        let sessionConfig = URLSessionConfiguration.ephemeral
        sessionConfig.connectionProxyDictionary = [
          "SOCKSEnable": 1,
          "SOCKSProxy": "127.0.0.1",
          "SOCKSPort": self.socksPort
        ] as [String: Any]

        sessionConfig.timeoutIntervalForRequest = 30
        sessionConfig.timeoutIntervalForResource = 30

        let session = URLSession(configuration: sessionConfig)

        // Create request
        var request = URLRequest(url: requestUrl)
        request.httpMethod = method

        // Add headers
        for (key, value) in headers {
          request.setValue(value, forHTTPHeaderField: key)
        }

        // Add body if present
        if let bodyString = body, ["POST", "PUT", "PATCH"].contains(method) {
          request.httpBody = bodyString.data(using: .utf8)
          if request.value(forHTTPHeaderField: "Content-Type") == nil {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
          }
        }

        // Execute request
        let task = session.dataTask(with: request) { data, response, error in
          if let error = error {
            DispatchQueue.main.async {
              promise.reject("HTTP_REQUEST_ERROR", error.localizedDescription)
            }
            return
          }

          guard let httpResponse = response as? HTTPURLResponse else {
            DispatchQueue.main.async {
              promise.reject("HTTP_REQUEST_ERROR", "Invalid response")
            }
            return
          }

          // Parse response
          let responseData = data != nil ? String(data: data!, encoding: .utf8) ?? "" : ""

          // Convert headers to dictionary
          var responseHeaders: [String: String] = [:]
          for (key, value) in httpResponse.allHeaderFields {
            if let keyString = key as? String, let valueString = value as? String {
              responseHeaders[keyString] = valueString
            }
          }

          // Get status text
          let statusText = HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)

          // Resolve with response
          DispatchQueue.main.async {
            promise.resolve([
              "status": httpResponse.statusCode,
              "statusText": statusText,
              "headers": responseHeaders,
              "data": responseData,
              "url": url
            ])
          }
        }

        task.resume()
      }
    }

    // MARK: - View Definition (for compatibility)
    View(ExpoTorView.self) {
      Prop("url") { (view: ExpoTorView, url: URL) in
        if view.webView.url != url {
          view.webView.load(URLRequest(url: url))
        }
      }
      Events("onLoad")
    }
  }

  // MARK: - Private Helper Methods

  /**
   * Initializes Tor configuration with proper directories and settings
   * Creates necessary directories in app's documents folder
   */
  private func initializeTorConfiguration() throws {
    // Get app support directory for Tor data
    guard let appSupportDir = FileManager.default.urls(
      for: .applicationSupportDirectory,
      in: .userDomainMask
    ).first else {
      throw NSError(domain: "ExpoTor", code: 2, userInfo: [
        NSLocalizedDescriptionKey: "Could not find application support directory"
      ])
    }

    // Create Tor directory structure
    let torDir = appSupportDir.appendingPathComponent("tor", isDirectory: true)
    let dataDir = torDir.appendingPathComponent("data", isDirectory: true)
    let cacheDir = torDir.appendingPathComponent("cache", isDirectory: true)

    // Create directories if they don't exist
    try FileManager.default.createDirectory(at: torDir, withIntermediateDirectories: true)
    try FileManager.default.createDirectory(at: dataDir, withIntermediateDirectories: true)
    try FileManager.default.createDirectory(at: cacheDir, withIntermediateDirectories: true)

    // Create Tor configuration
    let config = TorConfiguration()
    config.dataDirectory = dataDir
    config.controlSocket = torDir.appendingPathComponent("control.socket")

    // Configure SOCKS port (use 0 for auto-assign to avoid conflicts)
    config.socksPort = 0 // Will be dynamically assigned

    // Additional Tor options for better connectivity
    config.options = [
      "Log": "notice stdout",
      "GeoIPFile": Bundle.main.path(forResource: "geoip", ofType: nil) ?? "",
      "GeoIPv6File": Bundle.main.path(forResource: "geoip6", ofType: nil) ?? ""
    ]

    // Store configuration
    self.torConfiguration = config

    // The actual SOCKS port will be retrieved after connection
  }

  /**
   * Establishes control connection to Tor and monitors bootstrap progress
   * Sends events to JavaScript as Tor bootstraps
   */
  private func establishControlConnection(completion: @escaping (Bool) -> Void) {
    guard let config = self.torConfiguration,
          let controlSocket = config.controlSocket else {
      completion(false)
      return
    }

    // Retry logic for establishing connection
    let maxAttempts = 40 // 20 seconds total (500ms * 40)

    func attemptConnection(currentAttempt: Int) {
      let attempts = currentAttempt + 1

      // Create controller (TorController initializer is not optional in newer versions)
      let controller = TorController(socketURL: controlSocket)

      // Try to connect
      do {
        try controller.connect()

        // Authenticate (no password needed for socket)
        guard let dataDir = config.dataDirectory else {
          self.sendEvent("onTorError", ["error": "Data directory not configured"])
          completion(false)
          return
        }

        let cookiePath = dataDir.appendingPathComponent("control_auth_cookie")
        let cookie = try Data(contentsOf: cookiePath)

        controller.authenticate(with: cookie) { [weak self] success, error in
          guard let self = self else { return }

          if success {
            self.torController = controller

            // Get actual SOCKS port
            controller.getInfoForKeys(["net/listeners/socks"]) { response in
              if let socksInfo = response.first,
                 let portRange = socksInfo.range(of: ":(\\d+)", options: .regularExpression) {
                let portString = String(socksInfo[portRange]).replacingOccurrences(of: ":", with: "")
                self.socksPort = Int(portString) ?? 9050
              }
            }

            // Monitor bootstrap progress
            self.monitorBootstrapProgress()

            completion(true)
          } else {
            self.sendEvent("onTorError", ["error": error?.localizedDescription ?? "Authentication failed"])
            completion(false)
          }
        }
      } catch {
        if attempts < maxAttempts {
          // Wait and retry
          DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            attemptConnection(currentAttempt: attempts)
          }
        } else {
          self.sendEvent("onTorError", ["error": error.localizedDescription])
          completion(false)
        }
      }
    }

    // Start first attempt after short delay to let Tor initialize
    DispatchQueue.global().asyncAfter(deadline: .now() + 1.0) {
      attemptConnection(currentAttempt: 0)
    }
  }

  /**
   * Monitors Tor bootstrap progress and sends status updates
   * Updates connection status when bootstrap completes
   */
  private func monitorBootstrapProgress() {
    guard let controller = self.torController else { return }

    // Add observer for bootstrap events
    controller.addObserver(forStatusEvents: { (type, severity, action, arguments) -> Bool in

      // Check bootstrap progress
      if action == "BOOTSTRAP" {
        if let progress = arguments?["PROGRESS"] {
          self.currentStatus = "BOOTSTRAPPING: \(progress)%"
          self.sendEvent("onTorStatus", ["status": self.currentStatus])

          // Check if fully bootstrapped (100%)
          if progress == "100" {
            self.isConnected = true
            self.currentStatus = "CONNECTED"
            self.sendEvent("onTorStatus", ["status": "CONNECTED"])
            self.sendEvent("onTorConnected", ["connected": true])
          }
        }
      }

      // Return true to continue receiving events
      return true
    })

    // Get initial bootstrap status
    controller.getInfoForKeys(["status/bootstrap-phase"]) { [weak self] response in
      guard let self = self else { return }
      if let phase = response.first {
        self.sendEvent("onTorStatus", ["status": phase])
      }
    }
  }
}
